parsing/location.ml

```diff
@@ -292,90 +375,184 @@ let highlight_terminfo lb ppf locs =
-          let line_bol_pos' = pos + 1 in
-          if line_bol_pos <= leftmost_start
-          && leftmost_start < line_bol_pos' then begin
-            line_start := line;
-            line_start_pos := line_bol_pos
-          end;
-          if line_bol_pos <= rightmost_end
-          && rightmost_end < line_bol_pos' then begin
-            line_end := line;
-            line_end_pos := pos
-          end;
-          loop (pos + 1) (line + 1) line_bol_pos'
-        else
-          loop (pos + 1) line line_bol_pos
-      in
-      loop 0 0 0;
-      !line_start, !line_end, !line_start_pos, !line_end_pos
+(* [get_lines] must return the lines to highlight, given starting and ending
+   positions.
+
+   See [get_lines_from_current_input] below for an instantiation of [get_lines]
+   that reads from the current input.
+*)
+let highlight_quote ppf
+    ~(get_lines:
+        start_pos:position -> end_pos:position ->
+      (string (* line *) * int (* beginning of line *)) list)
+    ?(max_lines = 10)
+    highlight_tag
+    locs
+  =
+  let iset = IMapSet.of_intervals @@ Misc.Stdlib.List.filter_map (fun loc ->
+    let s, e = loc.loc_start, loc.loc_end in
+    if s.pos_cnum = -1 || e.pos_cnum = -1 then None
+    else Some ((s, s.pos_cnum), (e, e.pos_cnum - 1))
+  ) locs in
+  match IMapSet.extrema iset with
+  | None -> ()
+  | Some ((leftmost, _), (rightmost, _)) ->
+      let lines =
+        get_lines ~start_pos:leftmost ~end_pos:rightmost
+        |> List.map (fun (line, s) ->
+          let line_nb =
+            match IMapSet.find_in iset ~s ~e:(s + String.length line - 1) with
+            | None -> ""
+            | Some p -> string_of_int p.pos_lnum
+          in
+          (line, line_nb, s)
+        )
     in
     Format.fprintf ppf "@[<v>";
-    (* Print the input, highlighting the locations.
-       Indent by two spaces. *)
-    Format.fprintf ppf "  @[<v>";
-    if line_start = line_end then begin
-      (* single-line error *)
-      for pos = line_start_pos to line_end_pos - 1 do
-        Format.pp_print_char ppf (read_char pos)
-      done;
-      Format.fprintf ppf "@,";
-      for pos = line_start_pos to rightmost_end - 1 do
-        if is_highlighted pos then Format.pp_print_char ppf '^'
-        else Format.pp_print_char ppf ' '
-      done
-    end else begin
-      (* multi-line error *)
-      for pos = line_start_pos to line_end_pos - 1 do
-        let c = read_char pos in
-        if c = '\n' then Format.fprintf ppf "@,"
-        else if is_highlighted pos then Format.pp_print_char ppf c
-        else Format.pp_print_char ppf '.'
-      done
+    begin match lines with
+    | [] | [("", _, _)] -> ()
+    | [(line, line_nb, line_start_cnum)] ->
+        (* Single-line error *)
+        Format.fprintf ppf "%s | %s@," line_nb line;
+        Format.fprintf ppf "%*s   " (String.length line_nb) "";
+        for pos = line_start_cnum to rightmost.pos_cnum - 1 do
+          (* NB: The two cases below are mutually exclusive *)
```

Comment:
There are four cases below, which two do we mean? Maybe use `else if` if the tests are exclusive?

Reply:
Indeed, that wasn't clear. I used an `else if` as you suggest.